### ISSUE TITLE
Problem on calculating the entropy

### ISSUE THREAD

**Issue by hubaak** [Cecile-hi/Multimodal-Learning-with-Alternating-Unimodal-Adaptation#2]
Thanks for your exciting work in dealing with the problem of multimodal imbalance. However, I met some troubles when running the code.
In `main.py`, the code for calculating entropy is shown below:
```python
def calculate_entropy(output):
    probabilities = F.softmax(output, dim=0)
    # probabilities = F.softmax(output, dim=1)
    log_probabilities = torch.log(probabilities)
    entropy = -torch.sum(probabilities * log_probabilities)
    return entropy
```
The size of the output is [B, N] (B for batch and N for the numbers of categories). 
According to the Equation (8) in the paper, the correct code to get entropy should be:
```python
def calculate_entropy(output):
    # probabilities = F.softmax(output, dim=0)
    probabilities = F.softmax(output, dim=1) # Softmax through categories
    log_probabilities = torch.log(probabilities)
    entropy = -torch.sum(probabilities * log_probabilities, dim=-1) # If you don't add "dim=-1" then you sum up the entropy in the whole batch
    return entropy
``` 
What makes a difference here is entropy should be calculated for each sample, but not be summed up in a batch.

Oh! By the way, I think Equation (8) looks a little bit odd as 'm' shows up on the left side of the equation and serves as indices of the 'max' operation.

![Equation 8 in the paper](https://github.com/Cecile-hi/Multimodal-Learning-with-Alternating-Unimodal-Adaptation/assets/56573828/5e45d943-e806-4799-b56d-b510822d65b7)

**Comment by LittlePoolSpirit** [comment]
I think you're right. When I run -- dynamic, the effect is even worse than the fixed fusion method.

**Comment by Cecile-hi** [comment]
Hi @hubaak, yeah, it seems we may have some mistakes on entropy calculation, we will do some carefully check on it and make some correction. 

In addition, thank you for your kind suggestion for the eq.8. In this eq, we want to use "m" to represent modality, however I think you are right, it looks a bit ambiguous, I will discuss this will my Prof and make any essential correction in the camera-ready paper.

**Comment by Cecile-hi** [comment]
Hi @LittlePoolSpirit, I think decreasing the eval batch szie may make some effect (e.g. --batch_size 1).
For example, using the released ckpt
With Fixed weight:
```
python main.py --ckpt_path best_model_of_dataset_CREMAD_Normal_alpha_0.3_optimizer_sgd_modulate_starts_0_ends_50_epoch_91_acc_0.7768817204301075.pth --gs_flag --dataset CREMAD --batch_size 1 --lorb base --av_alpha 0.55
```
We have: Accuracy: 0.7768817204301075, accuracy_a: 0.5981182795698925, accuracy_v: 0.668010752688172

--dyanmic (with the modification by @hubaak )
```
python main.py --ckpt_path best_model_of_dataset_CREMAD_Normal_alpha_0.3_optimizer_sgd_modulate_starts_0_ends_50_epoch_91_acc_0.7768817204301075.pth --gs_flag --dataset CREMAD --batch_size 1 --lorb base --dynamic
``` 
We have: Accuracy: 0.7876344086021505, accuracy_a: 0.5981182795698925, accuracy_v: 0.668010752688172

**Comment by ggamaz** [comment]
i just find that the eq.8 can be simplifed as follow:
![image](https://github.com/user-attachments/assets/b0e2cc75-081c-4b35-8696-89f461e89724)
it just may be a softmax as follow:
![image](https://github.com/user-attachments/assets/d896b72a-de56-4b5b-9efd-3928dd36e62b)

**Comment by hubaak** [comment]
> i just find that the eq.8 can be simplifed as follow: ![image](https://private-user-images.githubusercontent.com/103712346/377707749-b0e2cc75-081c-4b35-8696-89f461e89724.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk1MTIzNTgsIm5iZiI6MTcyOTUxMjA1OCwicGF0aCI6Ii8xMDM3MTIzNDYvMzc3NzA3NzQ5LWIwZTJjYzc1LTA4MWMtNGIzNS04Njk2LTg5ZjQ2MWU4OTcyNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyMVQxMjAwNThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZjViNjJhMzk1N2IxMWQ0MGY0M2Q0MDA3YzNiZjM5MTIzYjFmOThlODAxYzcwN2NhMjBiOTYwMDJmMDM2ZTAzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.adV6J5akUVRFgVz0u0d8PUjEIygaWLCp6jp9ahBVO2w) it just may be a softmax as follow: ![image](https://private-user-images.githubusercontent.com/103712346/377709642-d896b72a-de56-4b5b-9efd-3928dd36e62b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk1MTIzNTgsIm5iZiI6MTcyOTUxMjA1OCwicGF0aCI6Ii8xMDM3MTIzNDYvMzc3NzA5NjQyLWQ4OTZiNzJhLWRlNTYtNGI1Yi05ZWZkLTM5MjhkZDM2ZTYyYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyMVQxMjAwNThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03Mjc2YjRhNmRiMDRiY2YzOTBjOGYyMDBkZWIyMzJmNzBjYmUzMWUwOTYzN2EwODkyZTE5M2ZjODA3ZDcwM2YxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.tSeyna3KlmAgucZ-uJk9d_WGPpHwKQxX28E5w-L1wfs)

Yeah, they simply weight the outputs of models with their minus entropy after softmax.

**Comment by Louise158** [comment]
Regarding the calculation of entropy, I think the following points need attention.
The paper mentions the formula for calculating entropy is as follows:
![image](https://github.com/user-attachments/assets/695712d5-3ddb-4556-885c-318fb56f9053)
Calculate the formula for the weights:
![image](https://github.com/user-attachments/assets/2011c79c-e12c-490c-827b-543e39873548)
It can be seen that this dynamic calculation of weights is performed on sample-level, which means calculating the weight for each sample's unimodal data. In this version of the code, the part responsible for calculating the weights is as follows:
![image](https://github.com/user-attachments/assets/d2200818-8bec-4e81-9ccc-97bba3a2f66a)
It can be seen that the calculation of weights is performed on batch-level here, which leads to poor inference performance.
Therefore, I think it can be modified like this:
![image](https://github.com/user-attachments/assets/03132b63-9a80-436c-ad82-31150186a47a)
